enhancements/implicit-catalog-versioning.md

```diff
@@ -0,0 +1,190 @@
+
+### Goals
+
+* Define Catalog update graphs for individual operators based implicitly on semantic version rather than asking users to explicitly mention what version the previous version replaces
+* Update OLM to handle the new format of catalogs
+* Handle conversion of existing operator subscriptions to use the new format
+
+### Non-Goals
+
+* Modify any of the features of OLM or the Operator Registry to use new information from the semantic version. This proposal simply suggests using the semantic version to define the graph
+* Resolve the migration of existing subscriptions to use the new Catalog Index images
+* Make any explicit changes to existing OLM API versions (CSV will remain `v1alpha1`)
+
+## Proposal
+
+Basic approach:
+
+- Make the `spec.version` field required going forward when building registry images with `opm registry` commands
+- Modify the operator registry to build dependency graph based simply on `spec.version` field rather than using a combination of channel heads and `spec.replaces` for each graph
+- Modify OLM to use the `spec.version` field everywhere it attempts to parse the graph for update information
+- Update skip feature to use version strings
+
+### Deprecate the `replaces` field:
+
+Currently, the update graph defined when creating an operator catalog is explicit. Essentially, the ClusterServiceVersion (CSV) file has an explicit `replaces` field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+    replaces: etcdoperator.v0.9.0
+```
+
+The `replaces` field is the source for the update graph for each install channel in a given operator. When the `operator-registry` builds the update graph, it actually walks through each CSV in a given Operator and, for each channel, constructs a chain of CSVs that define the update path. Instead, we can just define our spec without the replaces field:
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: etcdoperator.v0.9.2
+  namespace: placeholder
+spec:
+    displayName: etcd
+    description: Etcd Operator
+    version: 0.9.2
+```
+
+We would use the fact that 0.9.2 > 0.9.0 to implicitly assume that 0.9.0 is replaced by 0.9.2 through the rules of semantic version.
+
+To accomplish this:
+- Make `spec.version` a required field in the operator-registry project
+- Update the `operator-registry` to walk through the set of CSVs and construct the replacement chain by reading the `version` field instead of `replaces`
+- Add `version` as a key in the `operator-registry`'s bundle table to ensure that version has to be unique.
+- Add a `replacesVersion` field for each channel in the `operator-registry`'s database that defines which version this bundle is replacing in the update graph (along with a query to get the previous version)
```

Comment:
I need to review the schema to be sure, but I don' t think the steps from here down are needed. Channel entries can already be queried for the things listed, and I think the above steps cover generating channel entries from semver.

Reply:
I mentioned this in another comment, but I still think this separation is valuable for two purposes:

Firstly, just a simple separation of concerns. The operator-registry probably shouldn't have the implementation details of storing multiple different schema versions and responding differently depending on the contents. Otherwise as this schema changes over time it's going to probably get harder and harder to support lots of different ways to split data in different ways. Unless there's a really obvious design philosophy I'm missing, it seems much cleaner for specific registry versions to just only store the data they expect.

Secondly, this is also a really convenient way to cut support for old versions over time -- we probably don't want to support every version of every field forever. If we offer the new tool with the newest features and everyone moves onto that, at some point a version of OLM can include a support matrix that cuts support for some specific old version of registries over time.